packages/yarnpkg-fslib/sources/path.ts

```diff
@@ -81,30 +106,40 @@ export interface FormatInputPathObject<P extends Path> {
   name?: Filename;
 }
 
-export interface PathUtils<P extends Path> {
-  cwd(): P;
+export type MakeAbsolute<P> = P & (P extends PortablePath ? AbsolutePortablePath : AbsoluteNativePath);
+export type MakeRelative<P> = P & (P extends PortablePath ? RelativePortablePath : RelativeNativePath);
+
+export interface PathUtils<P extends Path, AP extends P = MakeAbsolute<P>, RP extends P = MakeRelative<P>> {
+  cwd(): AP;
 
-  normalize(p: P): P;
+  normalize<T extends P>(p: T): T;
+
+  join(initial: AP, ...paths: Array<P|Filename>): AP;
+  join(initial: RP | Filename, ...paths: Array<P|Filename>): RP;
   join(...paths: Array<P|Filename>): P;
-  resolve(...pathSegments: Array<P|Filename>): P;
-  isAbsolute(path: P): boolean;
-  relative(from: P, to: P): P;
-  dirname(p: P): P;
+  resolve(...pathSegments: Array<P|Filename>): AP;
```

Comment:
So that's the main problem I can see with this approach. The initial tweet was about a bug where a relative path was unexpectedly resolved relative to the cwd - but apart from that, the construct was technically fine per the `fs` prototypes: `resolve` is just fine with both relative and absolute paths.

So to fix the initial issue, we'd either have to type `Resolve` to forbid relative paths as first parameter (but that would be technically incorrect since the patched fs would still accept both relative and absolute paths), or make `xfs` a "stricter" version of `fs` where relative paths aren't allowed as paths.

What do you think about this?

Reply:
Either approach limits functionality via typescript while it works perfectly fine in javascript, which feels a bit weird.

That being said, it could make sense to have e.g. ZipFS, JailFS and/or CwdFS require relative paths, which would always be relative to the zip root/jail root/cwd. That way you can't pass in an absolute path, which could have accidentally been resolved with the process's working directory.

Reply:
Perhaps an option would be to make ZipFS & co generic on an "options" parameter:

```ts
declare const zipFs: ZipFS<NoRelative>;

declare const p: Path<Portable>;
declare const rel: Path<Portable & Relative>;
declare const abs: Path<Portable & Absolute>;

// Doesn't work
zipFs.readFileSync(p);
zipFs.readFileSync(rel);

// Works
zipFs.readFileSync(abs);
```

This option would default to false (allows relative), but `xfs` would default to `true`, for enforced strictness.